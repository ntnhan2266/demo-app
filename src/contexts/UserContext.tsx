import React, { createContext, useState, ReactNode } from 'react';
import { PublicUser } from '@/interfaces/user';

// Context interface
interface IUserContext {
  user: PublicUser | null;
  setUser: (_user: PublicUser | null) => void;
}

// Create the context
export const UserContext = createContext<IUserContext | undefined>(undefined);

// Provider component
export const UserProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [user, setUser] = useState<PublicUser | null>(null);

  const contextValue: IUserContext = {
    user,
    setUser,
  };

  return <UserContext.Provider value={contextValue}>{children}</UserContext.Provider>;
};
