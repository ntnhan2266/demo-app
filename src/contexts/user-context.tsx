import React, { createContext, useContext, useState, ReactNode } from 'react';
import { PublicUser } from '@/interfaces/user';

// Context interface
interface IUserContext {
  user: PublicUser | null;
  setUser: (user: PublicUser | null) => void;
}

// Create the context
const UserContext = createContext<IUserContext | undefined>(undefined);

// Provider component
export const UserProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
  const [user, setUser] = useState<PublicUser | null>(null);

  const contextValue: IUserContext = {
    user,
    setUser,
  };

  return <UserContext.Provider value={contextValue}>{children}</UserContext.Provider>;
};

// Custom hook to use the user context
export const useUser = () => {
  const context = useContext(UserContext);

  if (!context) {
    throw new Error('useUser must be used within a UserProvider');
  }

  return context;
};
