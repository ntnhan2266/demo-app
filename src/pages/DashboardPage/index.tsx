import React from 'react';
import { Route, Routes } from 'react-router-dom';
import Sidebar from '@/components/dashboard/Sidebar';
import UpdateUserInfoForm from '@/components/dashboard/UpdateUserInfoForm';
import GitHubRepos from '@/components/dashboard/GitHubRepos';

const DashboardPage: React.FC = () => {
  return (
    <div className="flex h-screen bg-gray-100">
      <Sidebar />

      <div className="flex-1 flex flex-col overflow-hidden">
        <main className="flex-1 overflow-x-hidden overflow-y-auto p-4">
          <Routes>
            <Route path='/update-info' element={<UpdateUserInfoForm />} />
            <Route path='/github-repos' element={<GitHubRepos />} />
          </Routes>
        </main>
      </div>
    </div>
  );
};

export default DashboardPage;
