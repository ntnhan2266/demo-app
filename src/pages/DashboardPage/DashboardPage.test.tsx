import React from 'react';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import DashboardPage from '@/pages/DashboardPage';
import { IUserContext, UserContext } from '@/contexts/UserContext';
import renderer from 'react-test-renderer';

// Mock the withAuthentication HOC
jest.mock('@/hocs/withAuthentication', () => (Component: React.FC) => Component);

describe('DashboardPage', () => {
  const contextValue: IUserContext = {
    user: {
      firstName: 'John',
      lastName: 'Doe',
      emailOrPhone: 'john.doe@example.com',
    },
    setUser: () => {},
  };

  test('should match the snapshot', () => {
    const tree = renderer
      .create(
        <UserContext.Provider value={contextValue}>
          <MemoryRouter initialEntries={['/github-repos']} initialIndex={0}>
            <DashboardPage />
          </MemoryRouter>
        </UserContext.Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders dashboard page with sidebar', () => {
    render(
      <UserContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={['/github-repos']} initialIndex={0}>
          <DashboardPage />
        </MemoryRouter>
      </UserContext.Provider>,
    );

    // Ensure that the dashboard page is rendered
    const dashboardPageElement = screen.getByTestId('dashboard-page');
    expect(dashboardPageElement).toBeInTheDocument();

    // Ensure that the sidebar is rendered
    const sidebarElement = screen.getByTestId('sidebar');
    expect(sidebarElement).toBeInTheDocument();
  });

  test('renders UpdateUserInfoForm for /update-info route', () => {
    render(
      <UserContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={['/update-info']} initialIndex={0}>
          <DashboardPage />
        </MemoryRouter>
      </UserContext.Provider>,
    );

    // Ensure that UpdateUserInfoForm is rendered for /update-info route
    const updateInfoFormElement = screen.getByTestId('update-user-info-form');
    expect(updateInfoFormElement).toBeInTheDocument();
  });

  test('renders UpdateUserPasswordForm for /update-password route', () => {
    render(
      <UserContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={['/update-password']} initialIndex={0}>
          <DashboardPage />
        </MemoryRouter>
      </UserContext.Provider>,
    );

    // Ensure that UpdateUserPasswordForm is rendered for /update-password route
    const updatePasswordFormElement = screen.getByTestId('update-user-password-form');
    expect(updatePasswordFormElement).toBeInTheDocument();
  });

  test('renders GitHubRepos for /github-repos route', () => {
    render(
      <UserContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={['/github-repos']} initialIndex={0}>
          <DashboardPage />
        </MemoryRouter>
      </UserContext.Provider>,
    );

    // Ensure that GitHubRepos is rendered for /github-repos route
    const githubReposElement = screen.getByTestId('github-repos');
    expect(githubReposElement).toBeInTheDocument();
  });
});
