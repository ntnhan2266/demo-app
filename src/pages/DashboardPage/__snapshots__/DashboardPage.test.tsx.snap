// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardPage should match the snapshot 1`] = `
<div
  className="flex h-screen bg-gray-100"
  data-testid="dashboard-page"
>
  <div
    className="
          lg:w-64 lg:block lg:shadow-md transition-transform duration-300 ease-in-out relative
          "
    data-testid="sidebar"
  >
    <div
      className="p-4"
    >
      <h2
        className="text-xl font-semibold mb-2 text-blue-800"
      >
        John
         
        Doe
      </h2>
      <p
        className="mb-4 text-gray-600"
      >
        john.doe@example.com
      </p>
      <nav>
        <ul>
          <li
            className="mb-2"
          >
            <a
              className=" block w-full text-left p-2 rounded-md transition duration-300 hover:bg-indigo-100"
              href="/dashboard/update-info"
              onClick={[Function]}
            >
              Update user info
            </a>
          </li>
          <li
            className="mb-2"
          >
            <a
              className=" block w-full text-left p-2 rounded-md transition duration-300 hover:bg-indigo-100"
              href="/dashboard/update-password"
              onClick={[Function]}
            >
              Update password
            </a>
          </li>
          <li
            className="mb-2"
          >
            <a
              className=" block w-full text-left p-2 rounded-md transition duration-300 hover:bg-indigo-100"
              href="/dashboard/github-repos"
              onClick={[Function]}
            >
              Github repos
            </a>
          </li>
          <li
            className="mb-2"
          >
            <button
              className="rounded-md focus:outline-none w-full text-white text-base px-4 py-2 bg-gray-500 hover:bg-gray-600"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              Logout
            </button>
          </li>
        </ul>
      </nav>
    </div>
  </div>
  <div
    className="flex-1 flex flex-col overflow-hidden"
  >
    <main
      className="flex-1 overflow-x-hidden overflow-y-auto pt-8 pl-8 pr-8 pb-16"
    >
      <div
        data-testid="github-repos"
      >
        <h1
          className="text-2xl font-bold mb-8 uppercase"
        >
          Github Repos
        </h1>
        <div
          className="flex items-center gap-4"
        >
          <div
            className=""
          >
            <input
              className="border rounded-md px-3 py-2 focus:outline-none transition duration-300 w-full md:w-96 max-w-full w-96 border-gray-400"
              disabled={false}
              onChange={[Function]}
              onKeyDown={[Function]}
              placeholder="Enter your Github username"
              type="text"
              value=""
            />
          </div>
          <button
            className="rounded-md focus:outline-none w-40 text-white text-base px-4 py-2 bg-indigo-500 hover:bg-indigo-600"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Load repos
          </button>
        </div>
      </div>
    </main>
  </div>
</div>
`;
